### PR TITLE
RFC: replace `ANY` with `@nospecialize` annotation

### PR DESCRIPTION
**Pull request by JeffBezanson**
Part of #11339.

This provides a much nicer and more general approach to specialization annotations, that stays out of the way of the type system.

TODO:
- [x] NEWS and doc updates
- [x] Add deprecation message
- [x] Replace uses of `ANY` in Base

### CODE REVIEW

File: base/exports.jl [hunk 1]
```diff
@@ -1285,6 +1285,7 @@ export
     @simd,
     @inline,
     @noinline,
+    @nospecialize,
```
Comment (tkelman) [inline base/exports.jl]: add to stdlib doc index

File: test/meta.jl [hunk 2]
```diff
@@ -144,3 +144,14 @@ baremodule B
 end
 @test B.x == 3
 @test B.M.x == 4
+
+# specialization annotations
+function _nospec_some_args(@nospecialize(x), y, @nospecialize z::Int)
```
Comment (tkelman) [inline test/meta.jl]: is it redundant here on the z arg that has a concrete type constraint?

Comment (JeffBezanson) [inline test/meta.jl]: Pretty much; it will just have no effect in that case.

**Comment by iamed2** [comment]
Is there any functional difference between this and `ANY`?

**Comment by JeffBezanson** [comment]
No. But it will allow additional functionality in the future, such as `@nospecialize x::AbstractArray`, where the type and specialization are de-coupled.

File: src/julia.h [hunk 3]
```diff
@@ -255,6 +255,7 @@ typedef struct _jl_method_t {
 
     int32_t nargs;
     int32_t called;  // bit flags: whether each of the first 8 arguments is called
+    int32_t nospec;  // bit flags: which arguments should not be specialized
```
Comment (tkelman) [inline src/julia.h]: how many arguments can this use before it breaks down? will there be a sane error or warning with 32 or more arguments?

**Comment by andyferris** [comment]
Very nice - I like this a lot!

File: src/julia.h (hunk 3 above)
Comment (JeffBezanson) [inline src/julia.h]: I'll add a warning.

File: src/method.c [hunk 4]
```diff
@@ -642,6 +667,18 @@ JL_DLLEXPORT void jl_method_def(jl_svec_t *argdata,
     jl_methtable_t *mt;
     jl_sym_t *name;
     jl_method_t *m = NULL;
+    size_t i, na = jl_svec_len(atypes);
+    int32_t nospec = 0;
+    for(i=1; i < na; i++) {
+        jl_value_t *ti = jl_svecref(atypes, i);
+        if (ti == jl_ANY_flag ||
+            (jl_is_vararg_type(ti) && jl_tparam0(jl_unwrap_unionall(ti)) == jl_ANY_flag)) {
+            jl_depwarn("`x::ANY` is deprecated, use `@nospecialize(x)` instead.",
```
Comment (tkelman) [inline src/method.c]: add note in deprecated.jl

File: test/broadcast.jl [hunk 5]
```diff
@@ -346,7 +346,7 @@ end
 let f17314 = x -> x < 0 ? false : x
     @test eltype(broadcast(f17314, 1:3)) === Int
     @test eltype(broadcast(f17314, -1:1)) === Integer
-    @test eltype(broadcast(f17314, Int[])) === Union{Bool,Int}
+    @test eltype(broadcast(f17314, Int[])) == Union{Bool,Int}
```
Comment (tkelman) [inline test/broadcast.jl]: was this a bug or?

Comment (JeffBezanson) [inline test/broadcast.jl]: It's a bug in the test. The type might be `Union{Bool,Int}` or `Union{Int,Bool}`, which have different structures that can be distinguished by `===`. The `show` method for unions sorts the types though, obscuring this. #22664 is about normalizing union types so this doesn't come up.

Comment (tkelman) [inline test/broadcast.jl]: and changing ANY to nospecialize triggers a different result more often somehow?

Comment (JeffBezanson) [inline test/broadcast.jl]: Yes, that seems to be the case. I'm not sure it's worth the time to track down why.

File: base/deprecated.jl [hunk 6]
```diff
@@ -1519,6 +1519,8 @@ function replace(s::AbstractString, pat, f, n::Integer)
     end
 end
 
+# ::ANY is deprecated in src/method.c
```
Comment (tkelman) [inline base/deprecated.jl]: and jl_ANY_flag should be removed everywhere else in src, right?

File: src/subtype.c [hunk 7]
```diff
@@ -790,8 +790,6 @@ static int forall_exists_equal(jl_value_t *x, jl_value_t *y, jl_stenv_t *e);
 // diagonal rule (record_var_occurrence).
 static int subtype(jl_value_t *x, jl_value_t *y, jl_stenv_t *e, int param)
 {
-    if (x == jl_ANY_flag) x = (jl_value_t*)jl_any_type;
-    if (y == jl_ANY_flag) y = (jl_value_t*)jl_any_type;
```
Comment (tkelman) [inline src/subtype.c]: should these wait until the deprecation is removed?

Comment (JeffBezanson) [inline src/subtype.c]: I modified the method definition code to remove `ANY` from signatures before the type system sees them, so this should be ok. Using `ANY` outside a method signature is not allowed.

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22666#issuecomment-313920964) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/13abfbf70c221b4ff6ddd17566f4d240e46f638b/86a362a_vs_e1def10/report.md). cc @jrevels

**Comment by timholy** [comment]
FYI there are a couple new ANYs in boot.jl.

File: src/jltypes.c [hunk 8]
```diff
@@ -1937,10 +1933,11 @@ void jl_init_types(void)
                             "invokes",
                             "nargs",
                             "called",
+                            "nospec",
```
Comment (vtjnash) [inline src/jltypes.c]: no abbreviations. perhaps call this `nospecialize`?

File: src/method.c [hunk 9]
```diff
@@ -645,6 +670,18 @@ JL_DLLEXPORT void jl_method_def(jl_svec_t *argdata,
     jl_methtable_t *mt;
     jl_sym_t *name;
     jl_method_t *m = NULL;
+    size_t i, na = jl_svec_len(atypes);
+    int32_t nospec = 0;
+    for(i=1; i < na; i++) {
```
Comment (vtjnash) [inline src/method.c]: formatting

File: src/method.c [hunk 10]
```diff
@@ -645,6 +670,18 @@ JL_DLLEXPORT void jl_method_def(jl_svec_t *argdata,
...
+        jl_value_t *ti = jl_svecref(atypes, i);
+        if (ti == jl_ANY_flag ||
+            (jl_is_vararg_type(ti) && jl_tparam0(jl_unwrap_unionall(ti)) == jl_ANY_flag)) {
+            jl_depwarn("`x::ANY` is deprecated, use `@nospecialize(x)` instead.",
+                       (jl_value_t*)jl_symbol("ANY"));
+            nospec |= (1 << (i - 1));
```
Comment (vtjnash) [inline src/method.c]: `if (i < 32)`

File: src/method.c [hunk 11]
```diff
@@ -434,6 +434,30 @@ static void jl_method_set_source(jl_method_t *m, jl_code_info_t *src)
                 set_lineno = 1;
             }
         }
+        else if (jl_is_expr(st) && ((jl_expr_t*)st)->head == meta_sym &&
+                 jl_expr_nargs(st) > 1 && jl_exprarg(st,0) == (jl_value_t*)nospecialize_sym) {
```
Comment (vtjnash) [inline src/method.c]: space after comma

File: src/method.c [hunk 12]
```diff
@@ -434,6 +434,30 @@ static void jl_method_set_source(jl_method_t *m, jl_code_info_t *src)
...
+            for(size_t j=1; j < jl_expr_nargs(st); j++) {
```
Comment (vtjnash) [inline src/method.c]: formatting

File: src/method.c [hunk 13]
```diff
@@ -434,6 +434,30 @@ static void jl_method_set_source(jl_method_t *m, jl_code_info_t *src)
...
+                jl_value_t *aj = jl_exprarg(st, j);
+                if (jl_is_slot(aj)) {
+                    int sn = (int)jl_slot_number(aj) - 2;
+                    if (sn >= 0) {  // @nospecialize on self is valid but currently ignored
+                        if (sn > (m->nargs - 2)) {
+                            jl_printf(JL_STDERR,
+                                      "WARNING: @nospecialize annotation applied to a non-argument.\n");
+                        }
+                        else if (sn >= sizeof(m->nospec)*8) {
```
Comment (vtjnash) [inline src/method.c]: spaces around binary operator

File: src/method.c [hunk 14]
```diff
@@ -434,6 +434,30 @@ static void jl_method_set_source(jl_method_t *m, jl_code_info_t *src)
...
+        else if (jl_is_expr(st) && ((jl_expr_t*)st)->head == meta_sym &&
+                 jl_expr_nargs(st) > 1 && jl_exprarg(st,0) == (jl_value_t*)nospecialize_sym) {
+            for(size_t j=1; j < jl_expr_nargs(st); j++) {
+                jl_value_t *aj = jl_exprarg(st, j);
+                if (jl_is_slot(aj)) {
+                    int sn = (int)jl_slot_number(aj) - 2;
+                    if (sn >= 0) {  // @nospecialize on self is valid but currently ignored
+                        if (sn > (m->nargs - 2)) {
+                            jl_printf(JL_STDERR,
+                                      "WARNING: @nospecialize annotation applied to a non-argument.\n");
```
Comment (vtjnash) [inline src/method.c]: make this an error

File: src/gf.c [hunk 15]
```diff
@@ -611,8 +611,8 @@ static void jl_cacheable_sig(
 
         int notcalled_func = (i > 0 && i <= 8 && !(definition->called & (1 << (i - 1))) &&
                               jl_subtype(elt, (jl_value_t*)jl_function_type));
-        if (decl_i == jl_ANY_flag) {
-            // don't specialize on slots marked ANY
+        if (i > 0 && i <= 8 && (definition->nospec & (1 << (i - 1))) &&
```
Comment (vtjnash) [inline src/gf.c]: `i <= sizeof(m->nospec) * 8`

File: src/ast.scm [hunk 16]
```diff
@@ -310,6 +318,9 @@
 (define (kwarg? e)
   (and (pair? e) (eq? (car e) 'kw)))
 
+(define (nospecialize-meta? e)
+  (and (length> e 2) (eq? (car e) 'meta) (eq? (cadr e) 'nospecialize)))
```
Comment (vtjnash) [inline src/ast.scm]: add another parameter `one::Bool` and check `(if one (eq? (length e) 3) (length> e 2))`

File: src/julia-syntax.scm [hunk 17]
```diff
@@ -3780,6 +3796,9 @@ f(x) = yt(x)
           ((and (pair? e) (eq? (car e) 'outerref))
            (let ((idx (get sp-table (cadr e) #f)))
                 (if idx `(static_parameter ,idx) (cadr e))))
+          ((and (length> e 2) (eq? (car e) 'meta) (eq? (cadr e) 'nospecialize))
```
Comment (vtjnash) [inline src/julia-syntax.scm]: `((nospecialize-meta e #f)`

**Review by vtjnash — COMMENTED** [review]
This looks good, just minor comments to fix.

**Comment by JeffBezanson** [comment]
Thanks. All comments addressed.

File: base/essentials.jl [hunk 18]
```diff
@@ -17,6 +17,42 @@ end
 macro _noinline_meta()
     Expr(:meta, :noinline)
 end
+
+"""
+    @nospecialize
```
Comment (tkelman) [inline base/essentials.jl]: I wonder whether it should be documented as `Base.@nospecialize` ?

File: base/inference.jl [hunk 19]
```diff
@@ -387,21 +387,21 @@ const t_ifunc_cost = Array{Int, 1}(n_ifunc)
...
 const t_ffunc_val = Array{Tuple{Int, Int, Any}, 1}(0)
 const t_ffunc_cost = Array{Int, 1}(0)
-function add_tfunc(f::IntrinsicFunction, minarg::Int, maxarg::Int, tfunc::ANY, cost::Int)
+function add_tfunc(f::IntrinsicFunction, minarg::Int, maxarg::Int, @nospecialize(tfunc), cost::Int)
     idx = reinterpret(Int32, f) + 1
     t_ifunc[idx] = (minarg, maxarg, tfunc)
     t_ifunc_cost[idx] = cost
 end
-function add_tfunc(#=@nospecialize::Builtin=# f::Function, minarg::Int, maxarg::Int, tfunc::ANY, cost::Int)
+function add_tfunc(#=@nospecialize::Builtin=# f::Function, minarg::Int, maxarg::Int, @nospecialize(tfunc), cost::Int)
```
Comment (tkelman) [inline base/inference.jl]: was the `#=@nospecialize::Builtin=#` left as a todo for this pr to address?

File: base/essentials.jl (hunk 18 above)
Comment (JeffBezanson) [inline base/essentials.jl]: Actually, not having it exported was pretty annoying. That's partly just an artifact of Base and tests using this more heavily than most code, but (1) other annotations like `@inline` are exported, and (2) it's hard to imagine `@nospecialize` conflicting with anything. So I'm more include to just export it.

File: base/boot.jl [hunk 20]
```diff
@@ -183,16 +183,23 @@ else
 end
 
 function Typeof end
-(f::typeof(Typeof))(x::ANY) = isa(x,Type) ? Type{x} : typeof(x)
+ccall(:jl_toplevel_eval_in, Any, (Any, Any),
+      Core, quote
+      (f::typeof(Typeof))(x) = ($(_expr(:meta,:nospecialize,:x)); isa(x,Type) ? Type{x} : typeof(x))
+      end)
```
Comment (vtjnash) [inline base/boot.jl]: why does this still need the more complex form? Also, why does it need `f::typeof(Typeof)` – could we just use `Typeof(@nospecialize x) = ...`

Comment (JeffBezanson) [inline base/boot.jl]: Because `Typeof` itself is used to lower method definitions not of the form `(f::T)(...) = ...`.

**Pull request merged**